src/annotator/index.ts

```diff
@@ -34,6 +34,36 @@ const sidebarLinkElement = document.querySelector(
   'link[type="application/annotator+html"][rel="sidebar"]',
 ) as HTMLLinkElement;
 
+/**
+ * Find and remove existing `<hypothesis-sidebar>` elements, and other
+ * Hypothesis application containers, which are created in the host frame.
+ *
+ * These might exist if the current page is a local snapshot of a web page saved
+ * with the browser's "Save Page As" feature. In that case the snapshot can
+ * include both the annotator bundle JS and the DOM elements it created. See
+ * https://github.com/hypothesis/client/issues/5827.
+ *
+ * Having duplicates of these elements is problematic because they contain
+ * iframed apps which will try to communicate with the host frame, and the
+ * host frame assumes there is only one of each.
+ *
+ * Returns true if any such elements were found.
+ */
+function removeExistingHypothesisAppElements(): boolean {
+  const appElements = document.querySelectorAll(
+    ['hypothesis-sidebar', 'hypothesis-notebook', 'hypothesis-profile'].join(
+      ',',
+    ),
+  );
+
+  let foundElements = false;
+  for (const appEl of appElements) {
+    appEl.remove();
+    foundElements = true;
+  }
+  return foundElements;
```

Comment:
Maybe we can return `appElements.length > 0` so that there's no need to define one more helper var.

We could even consider iterating with `appElements.forEach(...)`, instead of using a `for` loop.

```suggestion
  appElements.forEach(appEl => appEl.remove());
  return appElements.length > 0;
```

Reply:
Done. In an earlier iteration of this code I had some extra logic in the loop which excluded some elements from being removed.